django_irods/views.py

```diff
@@ -0,0 +1,477 @@
+    #     str(is_bag_download),
+    #     str(is_zip_download),
+    #     str(is_zip_request),
+    #     str(is_sf_agg_file)
+    # ))
+    # logger.debug(" after aggregation check, path={}, opath={}".format(
+    #     path, output_path
+    # ))
+
+    # determine active session
+    if res.is_federated:
+        # the resource is stored in federated zone
+        session = icommands.ACTIVE_SESSION
+    else:
+        # TODO: From Alva: I do not understand the use case for changing the environment.
+        # TODO: This seems an enormous potential vulnerability, as arguments are
+        # TODO: passed from the URI directly to IRODS without verification.
+        if 'environment' in kwargs:
+            logger.warn("setting iRODS from environment")
+            environment = int(kwargs['environment'])
+            environment = m.RodsEnvironment.objects.get(pk=environment)
+            session = Session("/tmp/django_irods", settings.IRODS_ICOMMANDS_PATH,
+                              session_id=uuid4())
+            session.create_environment(environment)
+            session.run('iinit', None, environment.auth)
+        elif getattr(settings, 'IRODS_GLOBAL_SESSION', False):
+            logger.debug("using GLOBAL_SESSION")
+            session = GLOBAL_SESSION
+        elif icommands.ACTIVE_SESSION:
+            logger.debug("using ACTIVE_SESSION")
+            session = icommands.ACTIVE_SESSION
+        else:
+            raise KeyError('settings must have IRODS_GLOBAL_SESSION set '
+                           'if there is no environment object')
+
+    resource_cls = check_resource_type(res.resource_type)
+
+    if is_zip_download:
+        pass  # zip already created; will be handled at the end; no further action necessary
+
+    elif is_zip_request or is_sf_agg_file:
+
+        if use_async:
+            # TODO: Why is there a wait of 3 seconds here? Changes so far have been synchronous!
+            task = create_temp_zip.apply_async((res_id, irods_path, irods_output_path,
+                                                is_sf_agg_file), countdown=3)
+            # TODO: 20 minutes might not be enough if the zipfile is large.
+            delete_zip.apply_async((irods_output_path, ),
+                                   countdown=(20 * 60))  # delete after 20 minutes
+
+            if rest_call:
+                return HttpResponse(
+                    json.dumps({
+                        'zip_status': 'Not ready',
+                        'task_id': task.task_id,
+                        'download_path': '/django_irods/rest_download/' + output_path}),
+                    content_type="application/json")
+            else:
+                # return status to the UI
+                request.session['task_id'] = task.task_id
+                # TODO: this is mistaken for a bag download in the UI!
+                # TODO: multiple asynchronous downloads don't stack!
```

Comment:
I don't quite understand the TODO comments above. Multiple asynchronous downloads will correspond to multiple request sessions, and each session will see its own async download to work. What did I miss?